File: 3510.minimum-pair-removal-to-sort-array-ii.java
#
# @lc app=leetcode id=3510 lang=java
#
# [3510] Minimum Pair Removal to Sort Array II
#

# @lc code=start
class Solution {
    public int minimumPairRemoval(int[] nums) {
        int n = nums.length;
        if (n < 2) return 0;
        Node[] nodes = new Node[n];
        for (int i = 0; i < n; i++) {
            nodes[i] = new Node((long) nums[i], i);
        }
        for (int i = 0; i < n - 1; i++) {
            nodes[i].next = nodes[i + 1];
            nodes[i + 1].prev = nodes[i];
        }
        java.util.PriorityQueue<Candidate> pq = new java.util.PriorityQueue<>();
        for (int i = 0; i < n - 1; i++) {
            pq.offer(new Candidate(nodes[i], nodes[i].val + nodes[i].next.val));
        }
        int violations = 0;
        for (int i = 0; i < n - 1; i++) {
            if (nums[i] > nums[i + 1]) violations++;
        }
        int ops = 0;
        while (violations > 0 && !pq.isEmpty()) {
            Candidate cand = pq.poll();
            Node left = cand.left;
            if (!left.active || left.next == null || !left.next.active) continue;
            long currSum = left.val + left.next.val;
            if (currSum != cand.sum) continue;
            Node prevv = left.prev;
            Node right = left.next;
            Node nextt = right.next;
            boolean hadA = prevv != null && prevv.val > left.val;
            boolean hadB = left.val > right.val;
            boolean hadC = nextt != null && right.val > nextt.val;
            violations -= (hadA ? 1 : 0) + (hadB ? 1 : 0) + (hadC ? 1 : 0);
            left.val += right.val;
            left.startIdx = Math.min(left.startIdx, right.startIdx);
            left.next = nextt;
            if (nextt != null) nextt.prev = left;
            right.active = false;
            boolean newA = prevv != null && prevv.val > left.val;
            boolean newC = nextt != null && left.val > nextt.val;
            violations += (newA ? 1 : 0) + (newC ? 1 : 0);
            ops++;
            if (prevv != null) {
                pq.offer(new Candidate(prevv, prevv.val + left.val));
            }
            if (left.next != null) {
                pq.offer(new Candidate(left, left.val + left.next.val));
            }
        }
        return ops;
    }

    private static class Node {
        long val;
        Node prev, next;
        int startIdx;
        boolean active = true;
        Node(long v, int idx) {
            val = v;
            startIdx = idx;
        }
    }

    private static class Candidate implements Comparable<Candidate> {
        long sum;
        Node left;
        Candidate(Node l, long s) {
            left = l;
            sum = s;
        }
        public int compareTo(Candidate o) {
            if (sum != o.sum) {
                return Long.compare(sum, o.sum);
            }
            return Integer.compare(left.startIdx, o.left.startIdx);
        }
    }
}
# @lc code=end

File: 3510.minimum-pair-removal-to-sort-array-ii.go
#
# @lc app=leetcode id=3510 lang=golang
#
# [3510] Minimum Pair Removal to Sort Array II
#

# @lc code=start
import "container/heap"

type Node struct {
	val  int64
	prev int
	next int
	pos  int
}

type Entry struct {
	sum  int64
	pos  int
	left int
}

type PriorityQueue []*Entry

func (pq PriorityQueue) Len() int {
	return len(pq)
}

func (pq PriorityQueue) Less(i, j int) bool {
	if pq[i].sum != pq[j].sum {
		return pq[i].sum < pq[j].sum
	}
	if pq[i].pos != pq[j].pos {
		return pq[i].pos < pq[j].pos
	}
	return pq[i].left < pq[j].left
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
}

func (pq *PriorityQueue) Push(x interface{}) {
	*pq = append(*pq, x.(*Entry))
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	*pq = old[:n-1]
	return item
}

func minimumPairRemoval(nums []int) int {
	n := len(nums)
	if n <= 1 {
		return 0
	}
	nodes := make([]Node, n)
	deleted := make([]bool, n)
	for i := 0; i < n; i++ {
		nodes[i].val = int64(nums[i])
		nodes[i].pos = i
		nodes[i].prev = -1
		nodes[i].next = -1
	}
	for i := 0; i < n-1; i++ {
		nodes[i].next = i + 1
		nodes[i+1].prev = i
	}
	pq := &PriorityQueue{}
	heap.Init(pq)
	bad := 0
	for i := 0; i < n-1; i++ {
		r := nodes[i].next
		if nodes[i].val > nodes[r].val {
			bad++
		}
		sumv := nodes[i].val + nodes[r].val
		heap.Push(pq, &Entry{sumv, nodes[i].pos, i})
	}
	currentLen := n
	ops := 0
	for bad > 0 && currentLen > 1 {
		var entry *Entry
		for pq.Len() > 0 {
			e := heap.Pop(pq).(*Entry)
			l := e.left
			if deleted[l] {
				continue
			}
			r := nodes[l].next
			if r == -1 || deleted[r] || nodes[r].prev != l {
				continue
			}
			if nodes[l].val + nodes[r].val != e.sum {
				continue
			}
			entry = e
			break
		}
		if entry == nil {
			break
		}
		l := entry.left
		r := nodes[l].next
		// subtract old bads
		if nodes[l].val > nodes[r].val {
			bad--
		}
		p := nodes[l].prev
		if p != -1 {
			if nodes[p].val > nodes[l].val {
				bad--
			}
		}
		nn := nodes[r].next
		if nn != -1 {
			if nodes[r].val > nodes[nn].val {
				bad--
			}
		}
		// merge r into l
		nodes[l].val += nodes[r].val
		deleted[r] = true
		nodes[l].next = nn
		if nn != -1 {
			nodes[nn].prev = l
		}
		currentLen--
		ops++
		// add new bads
		if p != -1 {
			if nodes[p].val > nodes[l].val {
				bad++
			}
		}
		if nn != -1 {
			if nodes[l].val > nodes[nn].val {
				bad++
			}
		}
		// push new pairs if exist
		if p != -1 {
			newsum := nodes[p].val + nodes[l].val
			heap.Push(pq, &Entry{newsum, nodes[p].pos, p})
		}
		if nn != -1 {
			newsum := nodes[l].val + nodes[nn].val
			heap.Push(pq, &Entry{newsum, nodes[l].pos, l})
		}
	}
	return ops
}
# @lc code=end